api/common/sharing/utils.go

```diff
@@ -10,46 +10,224 @@ import (
 	"github.com/Peripli/service-manager/pkg/web"
 	"github.com/Peripli/service-manager/storage"
 	"github.com/tidwall/gjson"
+	"reflect"
 )
 
-func ExtractReferencedInstanceID(ctx context.Context, repository storage.Repository, body []byte, tenantIdentifier string, getTenantId func() string) (string, error) {
+func ExtractReferencedInstanceID(req *web.Request, repository storage.Repository, body []byte, tenantIdentifier string, getTenantId func() string, smaap bool) (string, error) {
+	var err error
+	ctx := req.Context()
 	parameters := gjson.GetBytes(body, "parameters").Map()
+
+	err = validateParameters(parameters)
+	if err != nil {
+		return "", err
+	}
+
+	sharedInstances, err := getSharedInstancesByTenant(ctx, repository, tenantIdentifier, getTenantId())
+	if err != nil {
+		return "", err
+	}
+
+	var referencedInstance *types.ServiceInstance
 	referencedInstanceID, exists := parameters[instance_sharing.ReferencedInstanceIDKey]
+	if exists && referencedInstanceID.String() != "*" {
+		referencedInstance, err = getReferencedInstancesByID(ctx, repository, referencedInstanceID.String(), tenantIdentifier, getTenantId())
+		if err != nil {
+			return "", err
+		}
+	} else {
+		filteredInstancesList, err := filterInstancesBySelectors(ctx, repository, sharedInstances, parameters, smaap)
+		err = validateSingleResult(filteredInstancesList, parameters)
+		if err != nil {
+			return "", err
+		}
+		referencedInstance = filteredInstancesList.ItemAt(0).(*types.ServiceInstance)
+	}
 
-	if !exists {
-		return "", util.HandleInstanceSharingError(util.ErrMissingOrInvalidReferenceParameter, instance_sharing.ReferencedInstanceIDKey)
+	return referencedInstance.ID, nil
+}
+
+func IsValidReferenceInstancePatchRequest(req *web.Request, instance *types.ServiceInstance, planIDProperty string) error {
+	newPlanID := gjson.GetBytes(req.Body, planIDProperty).String()
+	if instance.ServicePlanID != newPlanID {
+		return util.HandleInstanceSharingError(util.ErrChangingPlanOfReferenceInstance, instance.ID)
 	}
+	parametersRaw := gjson.GetBytes(req.Body, "parameters").Raw
+	if parametersRaw != "" {
+		return util.HandleInstanceSharingError(util.ErrChangingParametersOfReferenceInstance, instance.ID)
+	}
+	return nil
+}
 
+func getReferencedInstancesByID(ctx context.Context, repository storage.Repository, referencedInstanceID string, tenantIdentifier string, tenantValue string) (*types.ServiceInstance, error) {
 	referencedInstanceObj, err := repository.Get(ctx, types.ServiceInstanceType,
-		query.ByLabel(query.EqualsOperator, tenantIdentifier, getTenantId()),
-		query.ByField(query.EqualsOperator, "id", referencedInstanceID.String()),
+		query.ByLabel(query.EqualsOperator, tenantIdentifier, tenantValue),
+		query.ByField(query.EqualsOperator, "id", referencedInstanceID),
```

Comment:
I think we do need it, we have it in the context but it is not good for us.
We have checked it before for the instance sharing feature.

Also, this line covers the flow of the container.

We can't use the criteria from context for the container flow because it also has the `container_id` key.
So when using the criteria from context for instances in a container, you actually filter by 2 keys.
We want to allow instance sharing for containers so need just the `tenantIdentifier`.